x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -141,4 +195,125 @@ public void removeFromCache(final Predicate<CacheKey> predicate) {
+    boolean isCacheFileToSync(CacheFile cacheFile) {
+        return cacheFilesToSync.contains(cacheFile);
+    }
+
+    /**
+     * Synchronize the cache files and their parent directories on disk.
+     *
+     * This method synchronizes the cache files that have been updated since the last time the method was invoked. To be able to do this,
+     * the cache files must notify the {@link CacheService} when they need to be fsync. When a {@link CacheFile} notifies the service the
+     * {@link CacheFile} instance is added to the current queue of cache files to synchronize referenced by {@link #cacheFilesToSync}.
+     *
+     * Cache files are serially synchronized using the {@link CacheFile#fsync()} method. When the {@link CacheFile#fsync()} call returns a
+     * non empty set of completed ranges this method also fsync the shard's snapshot cache directory, which is the parent directory of the
+     * cache entry. Note that cache files might be evicted during the synchronization.
+     */
+    protected void synchronizeCache() {
+        long count = 0L;
+        final Set<Path> cacheDirs = new HashSet<>();
+        final long startTimeNanos = threadPool.relativeTimeInNanos();
+        final int maxCacheFilesToSync = this.maxCacheFilesToSyncAtOnce;
+        for (long i = 0L; i < maxCacheFilesToSync; i++) {
+            if (lifecycleState() != Lifecycle.State.STARTED) {
+                logger.debug("stopping cache synchronization (cache service is closing)");
+                break;
+            }
+            final CacheFile cacheFile = cacheFilesToSync.poll();
+            if (cacheFile == null) {
+                logger.debug("stopping cache synchronization (no more cache files to fsync)");
+                break;
+            }
+            final Path cacheFilePath = cacheFile.getFile();
+            try {
+                final SortedSet<Tuple<Long, Long>> ranges = cacheFile.fsync();
+                if (ranges.isEmpty() == false) {
+                    logger.trace("cache file [{}] synchronized with [{}] completed range(s)", cacheFilePath.getFileName(), ranges.size());
+                    final Path cacheDir = cacheFilePath.toAbsolutePath().getParent();
+                    if (cacheDirs.add(cacheDir)) {
+                        try {
+                            IOUtils.fsync(cacheDir, true, false);
+                            logger.trace("cache directory [{}] synchronized", cacheDir);
+                        } catch (Exception e) {
+                            logger.warn(() -> new ParameterizedMessage("failed to synchronize cache directory [{}]", cacheDir), e);
+                        }
+                    }
+                    // TODO Index searchable snapshot shard information + cache file ranges in Lucene
+                    count += 1L;
+                }
+            } catch (Exception e) {
+                logger.warn(() -> new ParameterizedMessage("failed to fsync cache file [{}]", cacheFilePath.getFileName()), e);
```

Comment:
I agree, my attention was to not add the cache file to the Lucene index if the fsync failed.

>  Should we have a notion of forcefully dropping such a file from the cache since we can't trust it any longer?

I think we need such a mechanism because we should also discard a cache file in case we fail to write a range in it. I think I can try to tackle this in a follow up but I expect tests to be quite complex.